openforcefield/tests/test_molecule.py

```diff
@@ -529,6 +529,109 @@ def test_name(self):
         molecule.name = name
         assert molecule.name == name
 
+    def test_hill_formula(self):
+        """Test that making the hill formula is consistent between input methods and ordering"""
+        # make sure smiles match reference
+        molecule_smiles = Molecule.from_smiles('CCO')
+        assert molecule_smiles.hill_formula == 'C2H6O1'
+        # make sure is not order dependent
+        molecule_smiles_reverse = Molecule.from_smiles('OCC')
+        assert molecule_smiles.hill_formula == molecule_smiles_reverse.hill_formula
+        # make sure files and smiles match
+        molecule_file = Molecule.from_file(get_data_file_path('molecules/ethanol.mol2'))
+        assert molecule_smiles.hill_formula == molecule_file.hill_formula
+        # make sure the topology molecule gives the same formula
+        from openforcefield.topology.topology import TopologyMolecule, Topology
+        topology = Topology.from_molecules(molecule_smiles)
+        topmol = TopologyMolecule(molecule_smiles, topology)
+        assert molecule_smiles.hill_formula == Molecule.to_hill_formula(topmol)
+        # make sure the networkx matches
+        assert molecule_smiles.hill_formula == Molecule.to_hill_formula(molecule_smiles.to_networkx())
+
+    def test_isomorphic(self):
+        """Test the different levels of isomorphic matching"""
+        # check that hill formula fails are caught
+        ethanol = Molecule.from_smiles('CCO')
+        acetaldehyde = Molecule.from_smiles('CC=O')
+        assert ethanol.is_isomorphic_with(acetaldehyde) is False
+        assert acetaldehyde.is_isomorphic_with(ethanol) is False
+        # check that different orderings work with full matching
+        ethanol_reverse = Molecule.from_smiles('OCC')
+        assert ethanol.is_isomorphic_with(ethanol_reverse) is True
+        # check a reference mapping between ethanol and ethanol_reverse matches that calculated
+        ref_mapping = {0: 2, 1: 1, 2: 0, 3: 6, 4: 7, 5: 8, 6: 4, 7: 5, 8: 3}
+        assert Molecule.are_isomorphic(ethanol, ethanol_reverse, return_atom_map=True)[1] == ref_mapping
+        # check matching with nx.Graph atomic numbers and connectivity only
+        assert Molecule.are_isomorphic(ethanol, ethanol_reverse.to_networkx(), aromatic_matching=False,
+                                       formal_charge_matching=False, bond_order_matching=False,
+                                       atom_stereochemistry_matching=False,
+                                       bond_stereochemistry_matching=False)[0] is True
```

Comment:
This is a really complex function, but we should still test every permutation of these keyword arguments, for comparisons that evaluate either to `True` or `False`. Here's an example of how the toolkit tests to ensure that different permutations of nonbonded options produce the correct system: 

First it defines every permutation of inputs, and the expected output (or error message): https://github.com/openforcefield/openforcefield/blob/master/openforcefield/tests/test_forcefield.py#L318-L344

Then, it uses `pytest.parametrize` to run an array of individual tests: https://github.com/openforcefield/openforcefield/blob/master/openforcefield/tests/test_forcefield.py#L915-L930

These are all seen as separate tests by pytest, so each permutation of options either passes or fails, making debugging the separate possible causes much simpler. If the number of permutations becomes overwhelming, you can use itertools to define them all, and just define the expected behavior for the ethanol-vs-acetaldehyde comparison be `bond_order == True`